### PR TITLE
🌱 hack: add weekly update script for Slack

### PR DESCRIPTION
**Pull request by willie-yao**
**What this PR does / why we need it**:

This is a new PR based mainly off of the work by @oscr in #8192 
This PR adds a subcommand `weekly` to the existing `notes.go` command that generates the weekly updates that the comms team shares on Slack each week. This will make it simpler to create the messages and less chance of mistakes.

```
$ go run hack/tools/release/notes.go weekly --since 2023-08-13 --until 2023-08-20 --milestone v1.6

Weekly update :rotating_light:
Changes from 2023-08-13 to 2023-08-20 a total of 44 new commits where merged into main.

- 1 Breaking Changes :warning:
         - Improve Context handling in clusterctl (#8939)

- 5 New Features :sparkles:
         - Add automation to create release branch and tags (#9111)
         - block move with annotation (#8690)
         - machineset: adjust preflight check to allow kubelet version skew of 3 for clusters running v1.28 and above (#9222)
         - v1.28: Prepare quickstart, capd and tests for the new release including kind bump (#9160)
         - Add topology-owned label to MachineHealthChecks. (#9191)

- 5 Bug Fixes :bug:
         - test: pin conformance image to a version which includes a fix for the dualstack tests (#9252)
         - tilt: ensure .tiltbuild/bin directory is created early enough, add tilt troubleshooting guide (#9165)
         - Certificate paths in cloud-init scripts should not use a platform-dependent path separator (#9167)
         - Allow to drop useExperimentalRetryJoin field from KubeadmControlPlane.kubeadmConfigSpec (#9170)
         - Fix AddAnnotations for unstructured.Unstructured (#9164)

- 26 Other changes :seedling:

- 7 Documentation and book contributions :book: 

All merged PRs can be viewed in GitHub:
https://github.com/kubernetes-sigs/cluster-api/pulls?q=is%3Apr+closed%3A2023-08-13..2023-08-21+is%3Amerged+milestone%3Av1.6+

_Thanks to all our contributors!_ 😊
/Your friendly comms release team
```

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

/area release

### CODE REVIEW

**Review by killianmuldoon — COMMENTED** [review]
Could the `weekly` tool be a subcommand of the `release-notes` tool? I feel like this could become unwieldy and there's a duplication of effort. We might want another, similar, tool for emails for example.

**Comment by willie-yao** [comment]
> Could the `weekly` tool be a subcommand of the `release-notes` tool? I feel like this could become unwieldy and there's a duplication of effort. We might want another, similar, tool for emails for example.

I think that's a great idea actually. There does seem like a lot of duplicated code in the weekly script. I'll work on refactoring it into the notes script.

**Comment by willie-yao** [comment]
cc @kubernetes-sigs/cluster-api-release-team

File: hack/tools/release/notes.go
Comment (furkatgofurov7) [inline hack/tools/release/notes.go:478]: Thanks for the PR @willie-yao! I gave it a try running locally and something like:

```
go run hack/tools/release/notes.go weekly --since 2023-08-13 --until 2023-09-06 --milestone v1.6
```

outputs:

```
Weekly update :rotating_light:
Changes from 2023-08-13 to 2023-09-06 a total of 108 new commits where merged into main.

- 4 Breaking Changes :warning:
         - Remove v1alpha3 API Version (#8997)
         - : Remove go-vcs dependency from releaselink tool (#9288)
         - Bump to controller-runtime v0.16 (#8999)
         - Improve Context handling in clusterctl (#8939)

- 6 New Features :sparkles:
         - Add MachinePool workers support in ClusterClass (#9016)
         - Add automation to create release branch and tags (#9111)
         - block move with annotation (#8690)
         - machineset: adjust preflight check to allow kubelet version skew of 3 for clusters running v1.28 and above (#9222)
         - v1.28: Prepare quickstart, capd and tests for the new release including kind bump (#9160)
         - Add topology-owned label to MachineHealthChecks. (#9191)

- 12 Bug Fixes :bug:
         - Fix autoscaler image repo (#9353)
         - fix go install path for kpromo v4.0.4 (#9336)
         - MD controller: use regular random suffix for MachineSets, ensure max length 63 (#9298)
         - Deduplicating area in pr title in release notes (#9186)
         - Adding metrics container port in tilt-prepare only if it's missing (#9308)
         - Fix tilt-prepare leader-elect setting (#9315)
         - Allow duplicate objects in Tiltfile (#9302)
         - test: pin conformance image to a version which includes a fix for the dualstack tests (#9252)
         - tilt: ensure .tiltbuild/bin directory is created early enough, add tilt troubleshooting guide (#9165)
         - Certificate paths in cloud-init scripts should not use a platform-dependent path separator (#9167)
         - Allow to drop useExperimentalRetryJoin field from KubeadmControlPlane.kubeadmConfigSpec (#9170)
         - Fix AddAnnotations for unstructured.Unstructured (#9164)

- 76 Other changes :seedling:

- 10 Documentation and book contributions :book: 

All merged PRs can be viewed in GitHub:
https://github.com/kubernetes-sigs/cluster-api/pulls?q=is%3Apr+closed%3A2023-08-13..2023-09-07+is%3Amerged+milestone%3Av1.6+

_Thanks to all our contributors!_ 😊
/Your friendly comms release team
```

I wonder if we should also consider a case of PR titles like in https://github.com/kubernetes-sigs/cluster-api/pull/9288 PR in the code or just fix it manually 😄 

Anyways, to have a proper output for just :warning: PRs (it could happen to any PR category), this change:
```suggestion
		case strings.HasPrefix(body, ":warning:"), strings.HasPrefix(body, "⚠️"):
			key = warning
			parts := strings.Split(body, ":")
			if len(parts) > 1 {
				body = strings.TrimSpace(parts[len(parts)-1])
			} else {
				body = strings.TrimPrefix(body, ":warning:")
				body = strings.TrimPrefix(body, "⚠️")
				body = strings.TrimSpace(body)
			}
```

should suffice, but again, we might have more than `:` or any other unwanted/unpredictable char in the PR titles due to typos/by overlooking 🙂  So my intent here is, we either consider covering the corner uses cases like the above and extend the code a bit or just ignore it since it is not a big deal and always can be fixed manually.

**Review by furkatgofurov7 — COMMENTED** [review]
Great job! One comment inline:

File: hack/tools/release/notes.go
Comment (willie-yao) [reply hack/tools/release/notes.go:478]: Thanks for the suggestion @furkatgofurov7! I think we should be diligent to catch these errors and I feel like there may be too many edge cases to account for all possible typos. However I do agree that it would be nice to have it be fixed automatically. Wdyt @g-gaston @mjlshen @nprokopic?

File: hack/tools/release/notes.go
Comment (mjlshen) [inline hack/tools/release/notes.go:498]:
```suggestion
	fmt.Printf("Changes from %v a total of %d new commits were merged into main.\n\n", commitRange, len(commits))
```

Also, should we replace "main" with the `milestone` variable or otherwise detect which branch we're on?

File: hack/tools/release/notes.go
Comment (mjlshen) [reply hack/tools/release/notes.go:478]: Yeah, I could see something like ":warning: - PR title" being used as well. Not sure if any other characters make sense other than ":" and "-", but I'm ok with letting them fall through for now as I think the notes still make sense.

**Comment by killianmuldoon** [comment]
>/cc @killianmuldoon @sbueringer hey folks, could you please help reviewing this, since it is very helpful for comms team

I've got it on my list - but for most of the release team stuff I've been waiting to see /lgtm from members of the release team to be sure it's in a state the team itself is happy with before doing a final review.

**Review by mjlshen — APPROVED** [review]
/lgtm

File: hack/tools/release/notes.go
Comment (killianmuldoon) [inline hack/tools/release/notes.go:498]: Should there be a newline
```suggestion
	fmt.Printf("Changes from %v:\n A total of %d new commits were merged into milestone %s.\n\n", commitRange, len(commits), *milestone)
```

Reads better IMO - but feel fine to keep as is if you prefer.

File: hack/tools/release/notes.go
Comment (killianmuldoon) [inline hack/tools/release/notes.go:521]:
```suggestion
	fmt.Println("https://github.com/kubernetes-sigs/cluster-api/pulls?q=is%3Apr+closed%3A" + *since + ".." + lastDay + "+is%3Amerged+milestone%3A" + *milestone + "+\n")
```

This isn't portable to other projects. That's probably fine as this is only used by CAPI today, butit might be worth highlighting that somewhere

Comment (killianmuldoon) [reply hack/tools/release/notes.go:521]: You could also take the project URL as an arg, this might also help with the above TODO.

File: hack/tools/release/notes.go
Comment (killianmuldoon) [inline hack/tools/release/notes.go:496]: Could this be solved by just taking the URL - or project name and org - as args to the command? Maybe better to do that up front.

**Review by killianmuldoon — COMMENTED** [review]
A couple of ideas to improve the overall flow here - but I'm fine if you want to take these as follow-ups:


- Add some docs for this in the release docs - just a simple explanation of what it's for and an example run is good enough IMO.
- Add a `help` subcommand to make this behaviour discoverable. 
- Put the current behaviour under its own subcommand so we don't have confusing behaviour.

File: hack/tools/release/notes.go
Comment (killianmuldoon) [reply hack/tools/release/notes.go:532]: Did you add the TODO? I'm not sure what it's asking to turn into a link, but it seems closer to L557

File: hack/tools/release/notes.go
Comment (killianmuldoon) [inline hack/tools/release/notes.go:249]: Sorry for my lack of clarity here - let's have this match the `repo` variable. We can assume `github.com` at this point, the same way we do in the nodesCmd.

**Review by g-gaston — COMMENTED** [review]
Note: none of this is strictly related your work, @willie-yao. Thanks for owning this!

I know it was proposed as part of this PR, but I have some doubts about making this part of the release notes command, at least on its current form.

I wonder what the benefit is. I understand the "duplicated code" issue (although I might have a different opinion on how bad duplicated code is in this case). However, making them both part of the same command is not the only way to avoid duplicated code: refactor into smaller methods, pull common functionality into common code, find an abstraction that fits both usecases and model them after it, etc.

In fact I would argue that by looking at this PR one can see that it actually does not solve the problem. There is still a decent amount of duplicated code. We went from duplicated code but with two differentiated pieces that could evolve and grow independently to duplicated code in one single (more rigid) piece. IMO this makes the code more difficult to work with and more difficult to reason about. One clear example is the "init" code where we parse the flags.

I wonder if we need to take a step back and rethink these tools. Btw this sentiment doesn't uniquely come from this PR, it is also fed by how many improvements we are making to this area lately. For exampled, one quick/cheap change would be to use cobra, which although it means taking on a dependency, it would make having subcommands easier.

All that said, if everyone is onboard with this change on its current form, let's go for it :)

**Comment by killianmuldoon** [comment]
@g-gaston agreed with everything you said - I think there's a bit of a balance between over-engineering this stuff and getting things done for the release team. This isn't production code for CAPI, but it will become foundational for the release process over time. I think it's a good idea to have a rethink.

In the mean time this tool can be used as-is to create slack messages without being merged into the core repoo.

cobra is a good idea IMO - we already use it in `clusterctl` so we won't impact the go.mod. (I also just checked and it doesn't import [`viper` which as some potential issues.](#9181)

**Comment by furkatgofurov7** [comment]
@killianmuldoon @g-gaston @willie-yao are we all on same agreement on this one and ready to go ahead with it? I am in favor of rethinking the tool, but since this patch as is helps the RT to save some time/effort let's take it in and create a separate issue for redesigning the tool. WDYT?

**Comment by willie-yao** [comment]
I'm +1 on going ahead with it since it makes generating notes for previous releases faster, although I think others are leaning towards not merging this due to duplicated code

**Comment by g-gaston** [comment]
my 2c
I would vote to change this back to be a separate command, even with duplicated code.

And after this merges, we can have improvement tasks to do things like (by order of complexity):
- Identify common/duplicated parts and pull them into reusable code.
- Use cobra to have subcommands and easier flag parsing.
- Rethink the structure of the programs. How can we model them so we don't repeat efforts but we can keep the different usecases as decoupled as possible?
- Re-think these commands and see if there is a common abstraction that fits both and if it's worth the tradeoff.

This are just examples, I would deflect to the release team decide what should be done here and on what order 😄

**Comment by killianmuldoon** [comment]
I'm happy to have the release team decide on the structure they'd prefer here - no strong opinion from my side as long as the tools are well maintained and don't become dead code or difficult to change over time.

**Comment by furkatgofurov7** [comment]
During the RT meeting we as a team decided to change this back to be a separate command as suggested by @g-gaston and later work on improvement points in separate issues.

**Comment by furkatgofurov7** [comment]
@willie-yao 👋🏼 is this ready now for review?

**Review by g-gaston — COMMENTED** [review]
Since you are moving the notes tool, can you please update the release task doc to poi8nt the build command to the right place?

**Comment by willie-yao** [comment]
@furkatgofurov7 Yes it's ready! 
@g-gaston Just updated the docs, got lost in the revert

**Comment by g-gaston** [comment]
Need to update this too
https://github.com/kubernetes-sigs/cluster-api/blob/main/Makefile#L1102

File: docs/release/release-tasks.md
Comment (furkatgofurov7) [inline docs/release/release-tasks.md:326]: We could have a new Makefile target for it as `release-notes-tool`, i.e: 
`release-weekly-update` and use that here

**Review by furkatgofurov7 — COMMENTED** [review]
Thanks for working on this, my last suggestion on this one:

**Review by g-gaston — COMMENTED** [review]
/lgtm

**Review by furkatgofurov7 — COMMENTED** [review]
Thanks @willie-yao!

/lgtm

File: docs/release/release-tasks.md
Comment (furkatgofurov7) [inline docs/release/release-tasks.md:324]:
```suggestion
2. Build the release weekly update tools binary.
```

**Review by furkatgofurov7 — CHANGES_REQUESTED** [review]

File: docs/release/release-tasks.md
Comment (willie-yao) [reply docs/release/release-tasks.md]: Whoops good catch! Just fixed it.

**Review by furkatgofurov7 — COMMENTED** [review]
/lgtm

cc @killianmuldoon

File: Makefile
Comment (furkatgofurov7) [reply Makefile:1105]: @willie-yao hey! it would be great if you have time to rebase + address the small comment above to get this in soon, thanks!

File: docs/release/release-tasks.md
Comment (furkatgofurov7) [inline docs/release/release-tasks.md:326]:
```suggestion
   make release-weekly-update-tool
```

**Review by furkatgofurov7 — CHANGES_REQUESTED** [review]
Last nit from my side:

File: docs/release/release-tasks.md
Comment (willie-yao) [reply docs/release/release-tasks.md:326]: Sorry for missing this and thanks for catching it!

**Review by furkatgofurov7 — COMMENTED** [review]
/lgtm

cc @killianmuldoon

**Review by killianmuldoon — COMMENTED** [review]
/approve

Thanks for this - good to see the quick evolution of the release tools this cycle!

**Comment by killianmuldoon** [comment]
Like with the release notes tool it would be better IMO to only ever use this from main instead of trying to backport all changes.

**Pull request merged**
